### PR TITLE
API Feature Consistency Test, Part 2

### PR DESCRIPTION
**Pull request by bashbaug**
This is part two of the API feature consistency test.  Part 1 was added in #875.

This adds feature consistency tests for:

* Program Scope Global Variables
* Non-Uniform Work Groups
* Read-Write Images
* Creating 2D Images from Buffers
* Depth Images
* Device and Host Timer Synchronization
* Intermediate Language Programs
* Subgroups
* Program Initialization and Clean-Up Kernels
* 3D Image Writes

Other minor changes:

* Adds string decoding for several missing error codes.
* Adds a `test_assert_error` macro to simplify checking for specific non-error-related conditions.
* Updated the consistency check for `CL_QUEUE_SIZE` to require the error `CL_INVALID_COMMAND_QUEUE` when called with a host command queue.

### CODE REVIEW

File: test_common/harness/errorHelpers.h
Comment (bashbaug) [reply test_common/harness/errorHelpers.h:83]: I think the main advantage of the previous macros is that they print which error occurred (via `IGetErrorString( errCode )`) and not just that the error was not `CL_SUCCESS`.  I agree they could be unified if we choose to change this behavior.  I'd prefer to do this as a separate PR though :).

Comment (jrprice) [reply test_common/harness/errorHelpers.h:83]: Good point, I somehow missed that. No action needed.

File: test_conformance/api/test_api_consistency.cpp
Comment (bashbaug) [reply test_conformance/api/test_api_consistency.cpp]: Fixed, at least in this test.  My quick grep shows 550 other instances of `int error` in the CTS tests though.  Probably not all of them are CL errors, but I bet most of them are.

File: test_conformance/api/test_api_consistency.cpp
Comment (bashbaug) [reply test_conformance/api/test_api_consistency.cpp:492]: Yes, definitely, not sure how I missed this one.  Fixed.

**Comment by bashbaug** [comment]
Please hold off merging this for just a bit - I may need to make one more small change as a result of:

https://github.com/KhronosGroup/OpenCL-Docs/pull/418/files#r478742972

I've converted this PR to a draft until the discussion is resolved.  Thanks!

**Comment by bashbaug** [comment]
I've made one small tweak to this test to address an observation in the related spec changes:

https://github.com/KhronosGroup/OpenCL-Docs/pull/418#discussion_r478742972

I think this is ready to go again.  Hopefully that was the last needed change!

**Comment by Jeremy-Kemp** [comment]
I've not had time this week to review the changes, however I have run the test on our implementation and don't have any comments yet based on that.

File: test_conformance/api/test_api_consistency.cpp
Comment (Jeremy-Kemp) [inline test_conformance/api/test_api_consistency.cpp:894]: `CL_DEVICE_ILS_WITH_VERSION` may return an empty array, not an empty string.

I think this should be updated to follow a similar path to the extended versioning test, making sure that a returned `cl_name_version_khr` is empty.

**Review by Jeremy-Kemp — CHANGES_REQUESTED** [review]

File: test_conformance/api/test_api_consistency.cpp
Comment (Jeremy-Kemp) [reply test_conformance/api/test_api_consistency.cpp:894]: Actually, I think this highlights a problem in extended versioning, not this consistency test.

I'll go and check and then update here.

Comment (bashbaug) [reply test_conformance/api/test_api_consistency.cpp:894]: Cool, I do think this code is correct.  There are two cases to consider:

1. What does `CL_DEVICE_IL_VERSION` return?  This is the single string query added in OpenCL 2.1.  This string is either empty (`""`, indicating that no ILs are supported) or non-empty (ILs supported), but it should not return size equal to zero.

2. What does `CL_DEVICE_ILS_WITH_VERSION` return?  This is the new OpenCL 3.0 array query.  This query may return size equal to zero (no array elements, indicating that no ILs are supported) or a non-zero size (some array elements, indicating that ILs are supported).

The check for `ilVersion == ""` is checking for the first case, and the check for `sz == 0` is checking for the second case.

File: test_common/harness/errorHelpers.h
Comment (kpet) [inline test_common/harness/errorHelpers.h:92]: nitpick: not sure we should always add the `!` after the message.

File: test_conformance/api/test_api_consistency.cpp
Comment (kpet) [reply test_conformance/api/test_api_consistency.cpp:293]: Yeah, agree.

File: test_conformance/api/test_api_consistency.cpp
Comment (kpet) [inline test_conformance/api/test_api_consistency.cpp:426]: I think this should test for `error == CL_INVALID_OPERATION || error == CL_INVALID_MEM_OBJECT`

File: test_conformance/api/test_api_consistency.cpp
Comment (kpet) [inline test_conformance/api/test_api_consistency.cpp:474]: Probably best to be defensive and initialise the variable to a non-zero value.

File: test_conformance/api/test_api_consistency.cpp
Comment (kpet) [inline test_conformance/api/test_api_consistency.cpp:543]: Could leave the CL C version selection to the framework. That would be a bit more future proof.

File: test_conformance/api/test_api_consistency.cpp
Comment (kpet) [inline test_conformance/api/test_api_consistency.cpp:714]: These will never fire.  Given the messages, I'm guessing the conditions are incomplete.

**Comment by bashbaug** [comment]
I created https://github.com/KhronosGroup/OpenCL-CTS/issues/933 to track possible changes to the atomic capability queries.  I'd like to merge this PR without these changes, if possible, and I'll create a follow-on PR to fix this last issue.

File: test_conformance/api/test_api_consistency.cpp
Comment (kpet) [inline test_conformance/api/test_api_consistency.cpp:721]: Hmm, it's the presence of the extension that defines whether the feature is supported. I would change the message to say something along the lines of `cl_khr_image2d_from_buffer` is supported but an invalid value was returned for ..."

File: test_conformance/api/test_api_consistency.cpp
Comment (kpet) [inline test_conformance/api/test_api_consistency.cpp:696]: Probably should test for the extension string here instead.

File: test_conformance/api/test_api_consistency.cpp
Comment (kpet) [inline test_conformance/api/test_api_consistency.cpp:755]: See comment above

File: test_conformance/api/test_api_consistency.cpp
Comment (Jeremy-Kemp) [reply test_conformance/api/test_api_consistency.cpp:894]: Perhaps this is where the confusion is coming from (empty string handling).

What exactly do we say the empty string is? In C its just `\0` which has a length of 0. That would mean that `sz` *can* be 0, but that is asserted against above.

File: test_conformance/api/test_api_consistency.cpp
Comment (kpet) [inline test_conformance/api/test_api_consistency.cpp:818]: A detail: I'm wondering whether "report" wouldn't be better than "support" in the error messages. We're defining support as reporting the extension _and_ at least one format.

**Review by kpet — CHANGES_REQUESTED** [review]

File: test_conformance/api/test_api_consistency.cpp
Comment (Jeremy-Kemp) [reply test_conformance/api/test_api_consistency.cpp:894]: (For reference, we currently return a size of 1 and have it simply contain `\0`)

File: test_common/harness/errorHelpers.h
Comment (bashbaug) [reply test_common/harness/errorHelpers.h:92]: I don't have a strong opinion about this, but I'd prefer to make any changes in a separate PR.  This macro follows the same conventions as the other macros in this file.

File: test_conformance/api/test_api_consistency.cpp
Comment (bashbaug) [reply test_conformance/api/test_api_consistency.cpp:426]: Yeah, I used to have both of these cases, but I think this is correct with the latest spec updates.  See: https://github.com/KhronosGroup/OpenCL-Docs/pull/418#discussion_r478742972

File: test_conformance/api/test_api_consistency.cpp
Comment (bashbaug) [reply test_conformance/api/test_api_consistency.cpp:474]: How strongly do you feel about this?  I'm sure this isn't the only test that initializes integral quantities to zero, and we'll return an error if the query fails so I think this is already fairly defensive.

File: test_conformance/api/test_api_consistency.cpp
Comment (bashbaug) [reply test_conformance/api/test_api_consistency.cpp:543]: This wasn't an option when I created this PR.  Fixed!

File: test_conformance/api/test_api_consistency.cpp
Comment (bashbaug) [reply test_conformance/api/test_api_consistency.cpp:714]: No, these may fire, since we'll enter the conditional if _either_ of the conditions are true.  These checks are to ensure that if either condition is true then both conditions are true.

File: test_conformance/api/test_api_consistency.cpp
Comment (bashbaug) [reply test_conformance/api/test_api_consistency.cpp:721]: Following the logic in the spec appendix, the feature is supported / not supported based on the device queries, and the extension must be reported consistent with these queries.  I think this is six of one, half-dozen of the other though, and it really doesn't matter which one gets checked first, so long as everything is consistent.  Are any consistency checks missing?

File: test_conformance/api/test_api_consistency.cpp
Comment (bashbaug) [reply test_conformance/api/test_api_consistency.cpp:696]: See above.

File: test_conformance/api/test_api_consistency.cpp
Comment (bashbaug) [reply test_conformance/api/test_api_consistency.cpp:818]: I don't feel strongly about this.  Do you want this changed?  If so, just tell me what to change it to :).

**Comment by bashbaug** [comment]
I've rebased these changes to pick up the changes to the CTS kernel framework to use the latest supported OpenCL C version.  I've replied to the remaining review comments.  If there are any others that are must-fix please call them out, and please be specific what must be changed.  Thanks!

File: test_common/harness/errorHelpers.h
Comment (kpet) [reply test_common/harness/errorHelpers.h:97]: Yeah, let's not spend more time on this.

File: test_conformance/api/test_api_consistency.cpp
Comment (kpet) [reply test_conformance/api/test_api_consistency.cpp:426]: Interesting, I'd missed the latest developments. Assuming we require the function to always be present in the dispatch table, then this is fine. I don't think we say that explicitly but maybe it's "obvious" since we require the function to be present and have a given behaviour.

File: test_conformance/api/test_api_consistency.cpp
Comment (kpet) [reply test_conformance/api/test_api_consistency.cpp:474]: It's not the only one but this one is particularly bad since we're asserting that the value is 0 directly after the query. An incorrect implementation that doesn't return anything (doesn't write to memory) for `CL_DEVICE_GLOBAL_VARIABLE_PREFERRED_TOTAL_SIZE` will pass the check with `sz` initialised to 0. It wouldn't if `sz` was initialised to a different value. `SIZE_MAX`?

File: test_conformance/api/test_api_consistency.cpp
Comment (kpet) [reply test_conformance/api/test_api_consistency.cpp:713]: Yup, I read the code a bit too quickly. The second check's condition should probably be `imageBaseAddressAlignment == 0` though.

File: test_conformance/api/test_api_consistency.cpp
Comment (kpet) [reply test_conformance/api/test_api_consistency.cpp:720]: Well the spec appendix is not normative and my reading of the normative section says otherwise. Both `CL_DEVICE_IMAGE_PITCH_ALIGNMENT` and `CL_DEVICE_IMAGE_BASE_ADDRESS_ALIGNMENT` are specified to return 0 when the feature is not supported but only the text for `CL_DEVICE_EXTENSIONS` defines a logical equivalence between the presence of the extension string and support for the feature.

> Are any consistency checks missing?

Not AFAICT. The test will be correct once the comment on the assert above is fixed.

Since this is mostly a readability issue and not a terrible one, I'll resolve this.

File: test_conformance/api/test_api_consistency.cpp
Comment (kpet) [reply test_conformance/api/test_api_consistency.cpp:817]: I won't lose sleep over that :).

**Review by kpet — CHANGES_REQUESTED** [review]
The assert in the 2d from buffer test needs fixing, my other comments needn't gate this PR IMO.

File: test_conformance/api/test_api_consistency.cpp
Comment (Jeremy-Kemp) [reply test_conformance/api/test_api_consistency.cpp:893]: I believe the issue is this:

Any device that does not support `CL_DEVICE_IL_VERSION` will fill out `1` for `param_value_size_ret`. Subsequently, `param_value` will be 1 byte and just contain `\0`.

`get_device_il_version_string` calls `get_device_info_string` which constructs an `std::string` using this data, resulting in a string whose length is not 0. Therefore, the constructed string is not equivalent to an `std::string("")`, so `ilVersion == ""` evaluates to `false`.

If my understanding is correct and I'm not missing something obvious (please correct me if I'm wrong), then I think this can be avoided in one of two ways.

1. `get_device_info_string` can be special cased for empty strings
2. The comparison in this test can be changed from `ilVersion == ""` to `strcmp(ilVersion.c_str(), "") == 0`

My preference would be to go with 1. (possibly as a separate PR), as the empty string case may crop up in other tests in the future and is not specific to this test.

File: test_conformance/api/test_api_consistency.cpp
Comment (bashbaug) [reply test_conformance/api/test_api_consistency.cpp:474]: I'll change the initialization value to `SIZE_MAX`.

File: test_conformance/api/test_api_consistency.cpp
Comment (bashbaug) [reply test_conformance/api/test_api_consistency.cpp:713]: Yes it should be `imageBaseAddressAlignment` - fixed!

File: test_conformance/api/test_api_consistency.cpp
Comment (bashbaug) [reply test_conformance/api/test_api_consistency.cpp:893]: Ah, I see.  Yes, we are constructing our `std::string` with:

```c
    std::vector<char> info(size);

    if ((err = clGetDeviceInfo(device, param_name, size, info.data(), NULL))
        != CL_SUCCESS)
    {
        throw std::runtime_error("clGetDeviceInfo failed\n");
    }

    return std::string(info.begin(), info.end());
```

I think we could change this to either `std::string(info.begin(), --info.end());` or `std::string(info.data(), size - 1)`, and get the behavior we are looking for.  Any preference?  I'm leaning towards the latter.

Comment (Jeremy-Kemp) [reply test_conformance/api/test_api_consistency.cpp:893]: No strong preference, so the latter looks good to me. Thanks.

Would you like me to do that as part of a new PR, or just do it with this PR?

Comment (bashbaug) [reply test_conformance/api/test_api_consistency.cpp:893]: I'll take care of it.  Thanks for the quick reply!

**Comment by bashbaug** [comment]
Great thanks!  I think I've addressed the remaining must-fix issues - please take a look.

File: test_conformance/api/test_api_consistency.cpp
Comment (Jeremy-Kemp) [reply test_conformance/api/test_api_consistency.cpp:893]: Thanks @bashbaug, looks good.

**Review by kpet — APPROVED** [review]

**Review by Jeremy-Kemp — APPROVED** [review]
Thanks for your help @bashbaug .

File: test_conformance/api/test_api_consistency.cpp
Comment (gwawiork) [inline test_conformance/api/test_api_consistency.cpp:1007]: What do you think about adding condition to check if supports_cl_khr_subgroups == cl_true then ifp == cl_true ?

Comment (bashbaug) [reply test_conformance/api/test_api_consistency.cpp:1007]: I thought this check would have existed already, but it doesn't look like it does.

I don't mind adding it but would you be OK adding it in a separate PR?  I'm going to need to make a minor modification to this test for #933 anyhow.

Comment (gwawiork) [reply test_conformance/api/test_api_consistency.cpp:1007]: We have such verification in subgroups test https://github.com/KhronosGroup/OpenCL-CTS/blob/master/test_conformance/subgroups/test_ifp.cpp#L356 . 
Then maybe there is no need to add such verification here again.

**Review by gwawiork — APPROVED** [review]
LGTM

**Pull request merged**
